File: sdks/python/src/memoturn/gemini.py
"""Drop-in Gemini wrapper — records client.models.generate_content and
client.models.generate_content_stream calls as generations."""
from __future__ import annotations

from typing import Any, Optional

from ._stream import _ABANDONED, _RecordingStream, _get
from .client import Memoturn, Trace
from .decorator import get_client


def wrap_gemini(client: Any, memoturn: Optional[Memoturn] = None, *, trace: Optional[Trace] = None) -> Any:
    """Patch client.models.generate_content and .generate_content_stream to trace calls.
    Returns the same client. Unlike OpenAI/Anthropic, Gemini has no stream=True flag —
    streaming is a separate method that always returns an iterator of full
    GenerateContentResponse chunks, so both methods are patched independently."""
    mt = memoturn or get_client()
    _patch_generate_content(client, mt, trace)
    _patch_generate_content_stream(client, mt, trace)
    return client


def _config_items(config: Any) -> dict[str, Any]:
    """Best-effort dict view of config — a plain dict, a pydantic-like model
    (model_dump()), a SimpleNamespace, or None. Never raises."""
    if config is None:
        return {}
    if isinstance(config, dict):
        return dict(config)
    dump = getattr(config, "model_dump", None)
    if callable(dump):
        try:
            return dump(exclude_none=True)
        except Exception:  # noqa: BLE001
            pass
    return dict(getattr(config, "__dict__", {}) or {})


def _split_config_input(contents: Any, config: Any) -> tuple[Any, dict[str, Any]]:
    items = _config_items(config)
    system_instruction = items.pop("system_instruction", None) or items.pop("systemInstruction", None)
    input_val = {"systemInstruction": system_instruction, "contents": contents} if system_instruction is not None else contents
    return input_val, items


def _patch_generate_content(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    models = client.models
    original = models.generate_content

    def generate_content(*args: Any, **kwargs: Any) -> Any:
        t = trace or mt.trace(name="gemini.generateContent")
        input_val, model_parameters = _split_config_input(kwargs.get("contents"), kwargs.get("config"))
        gen = t.generation(name="gemini.generateContent", model=kwargs.get("model"), provider="gemini",
                            input=input_val, modelParameters=model_parameters)
        try:
            resp = original(*args, **kwargs)
            gen.end(output=_output(resp), usage=_map_usage(_get(resp, "usage_metadata")))
            return resp
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

    models.generate_content = generate_content  # type: ignore[assignment]


def _patch_generate_content_stream(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    models = client.models
    original = getattr(models, "generate_content_stream", None)
    if original is None:
        return

    def generate_content_stream(*args: Any, **kwargs: Any) -> Any:
        t = trace or mt.trace(name="gemini.generateContentStream")
        input_val, model_parameters = _split_config_input(kwargs.get("contents"), kwargs.get("config"))
        gen = t.generation(name="gemini.generateContentStream", model=kwargs.get("model"), provider="gemini",
                            input=input_val, modelParameters=model_parameters)
        try:
            raw = original(*args, **kwargs)
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

        acc = _GeminiAccumulator()

        def on_chunk(chunk: Any) -> None:
            acc.add(chunk)

        def on_done(err: Any) -> None:
            output, usage = acc.finalize()
            if err is _ABANDONED:
                gen.end(level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage)
            elif err is not None:
                gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
            else:
                gen.end(output=output, usage=usage)

        return _RecordingStream(raw, on_chunk, on_done)

    models.generate_content_stream = generate_content_stream  # type: ignore[assignment]


class _GeminiAccumulator:
    def __init__(self) -> None:
        self._text: Optional[str] = None
        self._usage: Any = None

    def add(self, chunk: Any) -> None:
        text = _get(chunk, "text")
        if text:
            self._text = (self._text or "") + text
        usage = _get(chunk, "usage_metadata")
        if usage is not None:
            self._usage = usage

    def finalize(self) -> tuple[Any, Optional[dict[str, Any]]]:
        return self._text, _map_usage(self._usage)


def _output(resp: Any) -> Any:
    text = _get(resp, "text")
    if text is not None:
        return text
    candidates = _get(resp, "candidates")
    if candidates is not None:
        try:
            return [c.model_dump() if hasattr(c, "model_dump") else c for c in candidates]
        except Exception:  # noqa: BLE001
            return str(candidates)
    return str(resp)


def _map_usage(usage: Any) -> Optional[dict[str, Any]]:
    if usage is None:
        return None
    prompt_tokens = _get(usage, "prompt_token_count")
    completion_tokens = _get(usage, "candidates_token_count")
    cache_read_tokens = _get(usage, "cached_content_token_count")
    out: dict[str, Any] = {"promptTokens": prompt_tokens, "completionTokens": completion_tokens}
    if prompt_tokens is not None and completion_tokens is not None:
        out["totalTokens"] = prompt_tokens + completion_tokens
    if cache_read_tokens is not None:
        out["cacheReadTokens"] = cache_read_tokens
    return out

File: sdks/js/src/gemini.ts
import type { Memoturn, MemoturnTrace } from "./client.js";
import { tapStream } from "./stream.js";

/**
 * Drop-in wrapper for the Google Gemini SDK (`@google/genai`). Wraps both
 * `client.models.generateContent` and `client.models.generateContentStream` so each call is
 * recorded as a memoturn generation (model, params, usage, latency, output).
 *
 * Structurally different from OpenAI/Anthropic: Gemini has **no `stream: true` flag** —
 * streaming is a completely separate, always-streaming method, wrapped independently from the
 * non-streaming one. Each yielded chunk is a *full* response object (the same shape as the
 * non-streaming return value), not a distinct delta type: `.text` per chunk is an incremental
 * delta (concatenated across chunks for `output`), while `.usageMetadata` is cumulative in
 * Gemini's protocol — the latest non-null one is taken as-is, never accumulated manually.
 *
 *   const gemini = wrapGemini(new GoogleGenAI({ apiKey }), memoturn);
 *   await gemini.models.generateContent({ model, contents, config });
 *   await gemini.models.generateContentStream({ model, contents, config });
 *
 * Pass `{ trace }` to nest generations under an existing trace; otherwise each call gets its
 * own trace. Pass `{ streamTimeoutMs }` to override the idle-stream abandonment backstop
 * (default 120s).
 */
export function wrapGemini<T extends object>(client: T, memoturn: Memoturn, options: WrapOptions = {}): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const value = Reflect.get(target, prop, receiver);
      if (prop === "models") return wrapModels(value, memoturn, options);
      return value;
    },
  });
}

type WrapOptions = { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number };

/** Split Gemini's nested `config` into the recorded `input` (systemInstruction promoted
 * alongside `contents`, mirroring Anthropic's `system` + `messages` combined input) and
 * `modelParameters` (everything else in `config` — OpenAI's exclusion philosophy, applied
 * to `config` instead of flat kwargs, since Gemini's config surface is large/unstable
 * rather than the small fixed set an allowlist fits). */
function splitConfigInput(
  contents: unknown,
  config: any,
): { input: unknown; modelParameters: Record<string, unknown> } {
  const { systemInstruction, ...modelParameters } = config ?? {};
  return {
    input: systemInstruction != null ? { systemInstruction, contents } : contents,
    modelParameters,
  };
}

function extractGeminiOutput(response: any): unknown {
  return response?.text ?? response?.candidates ?? response;
}

function mapGeminiUsage(usage: any): Record<string, number> | undefined {
  if (!usage) return undefined;
  const promptTokens = usage.promptTokenCount;
  const completionTokens = usage.candidatesTokenCount;
  return {
    promptTokens,
    completionTokens,
    ...(promptTokens != null && completionTokens != null ? { totalTokens: promptTokens + completionTokens } : {}),
    ...(usage.cachedContentTokenCount != null ? { cacheReadTokens: usage.cachedContentTokenCount } : {}),
  };
}

/** Merge streamed Gemini response chunks into the same `{ output, usage }` shape the
 * non-streaming path records. Each chunk is a full response object: `.text` is an
 * incremental delta (concatenated), `.usageMetadata` is cumulative (latest wins). */
function createGeminiStreamAccumulator() {
  let text = "";
  let sawText = false;
  let usage: any;
  return {
    add(chunk: any) {
      if (typeof chunk?.text === "string") {
        text += chunk.text;
        sawText = true;
      }
      if (chunk?.usageMetadata) usage = chunk.usageMetadata;
    },
    finalize() {
      return { output: sawText ? text : undefined, usage: mapGeminiUsage(usage) };
    },
  };
}

function wrapModels(models: any, memoturn: Memoturn, options: WrapOptions): any {
  return new Proxy(models, {
    get(target, prop, receiver) {
      const original = Reflect.get(target, prop, receiver);

      if (prop === "generateContent" && typeof original === "function") {
        return async function generateContent(params: any, ...rest: any[]) {
          const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "gemini.generateContent" });
          const { model, contents, config } = params ?? {};
          const { input, modelParameters } = splitConfigInput(contents, config);
          const generation = trace.generation({
            name: "gemini.generateContent",
            model,
            provider: "gemini",
            modelParameters,
            input,
          });
          try {
            const response = await original.call(target, params, ...rest);
            generation.end({
              output: extractGeminiOutput(response),
              usage: mapGeminiUsage(response?.usageMetadata),
            });
            return response;
          } catch (err) {
            generation.end({ level: "ERROR", statusMessage: String(err) });
            throw err;
          }
        };
      }

      if (prop === "generateContentStream" && typeof original === "function") {
        return async function generateContentStream(params: any, ...rest: any[]) {
          const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "gemini.generateContentStream" });
          const { model, contents, config } = params ?? {};
          const { input, modelParameters } = splitConfigInput(contents, config);
          const generation = trace.generation({
            name: "gemini.generateContentStream",
            model,
            provider: "gemini",
            modelParameters,
            input,
          });
          try {
            const response = await original.call(target, params, ...rest);
            const accumulator = createGeminiStreamAccumulator();
            return tapStream(
              response,
              {
                onChunk: (chunk: any) => accumulator.add(chunk),
                onDone: (err, reason) => {
                  if (reason === "error") {
                    generation.end({ level: "ERROR", statusMessage: String(err), ...accumulator.finalize() });
                  } else if (reason === "abandoned") {
                    generation.end({
                      level: "WARNING",
                      statusMessage: "stream ended before completion",
                      ...accumulator.finalize(),
                    });
                  } else {
                    generation.end({ ...accumulator.finalize() });
                  }
                },
              },
              { idleTimeoutMs: options.streamTimeoutMs },
            );
          } catch (err) {
            generation.end({ level: "ERROR", statusMessage: String(err) });
            throw err;
          }
        };
      }

      return original;
    },
  });
}
